auth/providers/azure/graph/graph.go

```diff
@@ -151,6 +158,111 @@ func (u *UserInfo) getExpandedGroups(ctx context.Context, ids []string) (*GroupL
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create getMemberGroups request")
+	}
+	// Set the auth headers
+	req.Header = u.headers
+
+	correlationID := uuid.New()
+	req.Header.Set("x-ms-correlation-request-id", correlationID.String())
+	getMemberGroupsCtx, cancel := context.WithTimeout(context.Background(), getMemberGroupsTimeout)
+	defer cancel()
+	klog.V(5).Infof("Sending getMemberGroups request with correlationID: %s", correlationID.String())
+	// use default httpclient without retries first
+	client := *httpclient.DefaultHTTPClient
+	resp, err := client.Do(req.WithContext(getMemberGroupsCtx))
+	if err != nil {
+		klog.V(5).Infof("CorrelationID: %s, Error: Failed to fetch group info using getMemberGroups: %s", correlationID.String(), err.Error())
+		return nil, errors.Errorf("CorrelationID: %s, Error: Failed to fetch group info using getMemberGroups", correlationID.String())
+	}
+
+	// use retryable client only for unavailable and gatewaytimeout errors
+	if resp.StatusCode == http.StatusServiceUnavailable || resp.StatusCode == http.StatusGatewayTimeout {
+		resp, err = u.client.Do(req.WithContext(ctx))
+		if err != nil {
+			klog.V(5).Infof("CorrelationID: %s, Error: Failed to fetch group info using getMemberGroups on retries: %s", correlationID.String(), err.Error())
+			return nil, errors.Errorf("CorrelationID: %s, Error: Failed to fetch group info using getMemberGroups on retries", correlationID.String())
+		}
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		data, _ := io.ReadAll(resp.Body)
+		klog.V(5).Infof("CorrelationID: %s, Error: Failed to fetch group info with status code: %d and response: %s", correlationID.String(), resp.StatusCode, string(data))
+		return nil, errors.Errorf("CorrelationID: %s, Error: Failed to fetch group info with status code: %d", correlationID.String(), resp.StatusCode)
+	}
+
+	groupResponse := struct {
+		Value []string `json:"value"`
+	}{}
+	// Decode the response
+	var groupNames []string
+	err = json.NewDecoder(resp.Body).Decode(&groupResponse)
+	if err != nil {
+		return nil, errors.Wrapf(err, "CorrelationID: %s, Error: Failed to decode response for request %s", correlationID.String(), req.URL.Path)
+	}
+
+	// Extract out the Group objects into a list of strings
+	for i := 0; i < len(groupResponse.Value); i++ {
+		groupNames = append(groupNames, groupResponse.Value[i])
+	}
+
+	return groupNames, nil
+}
+
+func getOBORegionalEndpointFunc(location string, resourceID string) (string, error) {
+	var suffix string
+
+	if strings.HasPrefix(location, "usgov") || strings.HasPrefix(location, "usdod") {
+		suffix = "us"
+	} else if strings.HasPrefix(location, "china") {
+		suffix = "cn"
+	} else {
+		suffix = "com"
+	}
+
+	if !strings.HasPrefix(resourceID, "/") {
+		resourceID = "/" + resourceID
+	}
+	return fmt.Sprintf(arcOboEndpointFormat, location, suffix, resourceID), nil
+}
+
 func (u *UserInfo) RefreshToken(ctx context.Context, token string) error {
```

Comment:
Should we refersh token only when it's expired?

Reply:
done